Battle of the neighborhoods New York vs Toronto

Description of the problem: 
In this work, two different sources of information will be handled, it seeks to work in order to segment and group the information of two locations such as Toronto, Canada and New York. In each of these, relevant information of this grouping of information by town, city, postal code will be displayed showing the different characteristics of the regions. All of the above with what was learned during the course.


Description of the data and how it will be used to solve the problem:
The data is obtained from web pages, once obtained, this information must be processed to be able to work it as a data frame, the data is cleaned to be able to filter it by city, neighborhood and postal code. Then some coordinates are obtained and the results are compared.


DATA ANALYSIS NEW YORK CITY

In [9]:
import numpy as np 

import pandas as pd 
pd.set_option('display.max_columns', None)
pd.set_option('display.max_rows', None)

import json 

from geopy.geocoders import Nominatim 

import requests 
from pandas.io.json import json_normalize


import matplotlib.cm as cm
import matplotlib.colors as colors


from sklearn.cluster import KMeans

print('Libraries imported.')



Libraries imported.


In [10]:
!wget -q -O 'newyork_data.json' https://cf-courses-data.s3.us.cloud-object-storage.appdomain.cloud/IBMDeveloperSkillsNetwork-DS0701EN-SkillsNetwork/labs/newyork_data.json
print('Data downloaded!')

Data downloaded!


In [12]:
with open('newyork_data.json') as json_data:
    newyork_data = json.load(json_data)

Explore the data

In [13]:
newyork_data

{'type': 'FeatureCollection',
 'totalFeatures': 306,
 'features': [{'type': 'Feature',
   'id': 'nyu_2451_34572.1',
   'geometry': {'type': 'Point',
    'coordinates': [-73.84720052054902, 40.89470517661]},
   'geometry_name': 'geom',
   'properties': {'name': 'Wakefield',
    'stacked': 1,
    'annoline1': 'Wakefield',
    'annoline2': None,
    'annoline3': None,
    'annoangle': 0.0,
    'borough': 'Bronx',
    'bbox': [-73.84720052054902,
     40.89470517661,
     -73.84720052054902,
     40.89470517661]}},
  {'type': 'Feature',
   'id': 'nyu_2451_34572.2',
   'geometry': {'type': 'Point',
    'coordinates': [-73.82993910812398, 40.87429419303012]},
   'geometry_name': 'geom',
   'properties': {'name': 'Co-op City',
    'stacked': 2,
    'annoline1': 'Co-op',
    'annoline2': 'City',
    'annoline3': None,
    'annoangle': 0.0,
    'borough': 'Bronx',
    'bbox': [-73.82993910812398,
     40.87429419303012,
     -73.82993910812398,
     40.87429419303012]}},
  {'type': 'Feature',
 

In [14]:
neighborhoods_data = newyork_data['features']

Show first item of the list

In [15]:
neighborhoods_data[0]

{'type': 'Feature',
 'id': 'nyu_2451_34572.1',
 'geometry': {'type': 'Point',
  'coordinates': [-73.84720052054902, 40.89470517661]},
 'geometry_name': 'geom',
 'properties': {'name': 'Wakefield',
  'stacked': 1,
  'annoline1': 'Wakefield',
  'annoline2': None,
  'annoline3': None,
  'annoangle': 0.0,
  'borough': 'Bronx',
  'bbox': [-73.84720052054902,
   40.89470517661,
   -73.84720052054902,
   40.89470517661]}}

Transform data into a pandas dataframe (Using columns like Borough, neighborhood, latitude y longitude).

In [16]:

column_names = ['Borough', 'Neighborhood', 'Latitude', 'Longitude'] 

neighborhoods = pd.DataFrame(columns=column_names)

In [17]:
neighborhoods

Empty DataFrame
Columns: [Borough, Neighborhood, Latitude, Longitude]
Index: []

In [18]:
for data in neighborhoods_data:
    borough = neighborhood_name = data['properties']['borough'] 
    neighborhood_name = data['properties']['name']
        
    neighborhood_latlon = data['geometry']['coordinates']
    neighborhood_lat = neighborhood_latlon[1]
    neighborhood_lon = neighborhood_latlon[0]
    
    neighborhoods = neighborhoods.append({'Borough': borough,
                                          'Neighborhood': neighborhood_name,
                                          'Latitude': neighborhood_lat,
                                          'Longitude': neighborhood_lon}, ignore_index=True)

You can see the information classified by Borough and Neighborhood 

In [19]:
neighborhoods.head()

Borough Neighborhood   Latitude  Longitude
0   Bronx    Wakefield  40.894705 -73.847201
1   Bronx   Co-op City  40.874294 -73.829939
2   Bronx  Eastchester  40.887556 -73.827806
3   Bronx    Fieldston  40.895437 -73.905643
4   Bronx    Riverdale  40.890834 -73.912585

it is found that there is a total of 5 boroughs and 306 neighborhoods

In [20]:
print('The dataframe has {} boroughs and {} neighborhoods.'.format(
        len(neighborhoods['Borough'].unique()),
        neighborhoods.shape[0]
    )
)

The dataframe has 5 boroughs and 306 neighborhoods.


We can see the geographic location of new york

In [21]:
address = 'New York City, NY'

geolocator = Nominatim(user_agent="ny_explorer")
location = geolocator.geocode(address)
latitude = location.latitude
longitude = location.longitude
print('The geograpical coordinate of New York City are {}, {}.'.format(latitude, longitude))

The geograpical coordinate of New York City are 40.7127281, -74.0060152.


In [ ]:
We can see all Neighborhoods

In [28]:
neighborhoods

Borough               Neighborhood   Latitude  Longitude
0            Bronx                  Wakefield  40.894705 -73.847201
1            Bronx                 Co-op City  40.874294 -73.829939
2            Bronx                Eastchester  40.887556 -73.827806
3            Bronx                  Fieldston  40.895437 -73.905643
4            Bronx                  Riverdale  40.890834 -73.912585
5            Bronx                Kingsbridge  40.881687 -73.902818
6        Manhattan                Marble Hill  40.876551 -73.910660
7            Bronx                   Woodlawn  40.898273 -73.867315
8            Bronx                    Norwood  40.877224 -73.879391
9            Bronx             Williamsbridge  40.881039 -73.857446
10           Bronx                 Baychester  40.866858 -73.835798
11           Bronx             Pelham Parkway  40.857413 -73.854756
12           Bronx                City Island  40.847247 -73.786488
13           Bronx               Bedford Park  40.870185 -73.885512
14           Bronx         University Heights  40.855727 -73.910416
15           Bronx             Morris Heights  40.847898 -73.919672
16           Bronx                    Fordham  40.860997 -73.896427
17           Bronx               East Tremont  40.842696 -73.887356
18           Bronx                 West Farms  40.839475 -73.877745
19           Bronx               High  Bridge  40.836623 -73.926102
20           Bronx                    Melrose  40.819754 -73.909422
21           Bronx                 Mott Haven  40.806239 -73.916100
22           Bronx                Port Morris  40.801664 -73.913221
23           Bronx                   Longwood  40.815099 -73.895788
24           Bronx                Hunts Point  40.809730 -73.883315
25           Bronx                 Morrisania  40.823592 -73.901506
26           Bronx                  Soundview  40.821012 -73.865746
27           Bronx               Clason Point  40.806551 -73.854144
28           Bronx                Throgs Neck  40.815109 -73.816350
29           Bronx               Country Club  40.844246 -73.824099
30           Bronx                Parkchester  40.837938 -73.856003
31           Bronx         Westchester Square  40.840619 -73.842194
32           Bronx                   Van Nest  40.843608 -73.866299
33           Bronx                Morris Park  40.847549 -73.850402
34           Bronx                    Belmont  40.857277 -73.888452
35           Bronx             Spuyten Duyvil  40.881395 -73.917190
36           Bronx            North Riverdale  40.908543 -73.904531
37           Bronx                 Pelham Bay  40.850641 -73.832074
38           Bronx              Schuylerville  40.826580 -73.826203
39           Bronx             Edgewater Park  40.821986 -73.813885
40           Bronx                Castle Hill  40.819014 -73.848027
41           Bronx                  Olinville  40.871371 -73.863324
42           Bronx             Pelham Gardens  40.862966 -73.841612
43           Bronx                  Concourse  40.834284 -73.915589
44           Bronx                  Unionport  40.829774 -73.850535
45           Bronx                   Edenwald  40.884561 -73.848083
46        Brooklyn                  Bay Ridge  40.625801 -74.030621
47        Brooklyn                Bensonhurst  40.611009 -73.995180
48        Brooklyn                Sunset Park  40.645103 -74.010316
49        Brooklyn                 Greenpoint  40.730201 -73.954241
50        Brooklyn                  Gravesend  40.595260 -73.973471
51        Brooklyn             Brighton Beach  40.576825 -73.965094
52        Brooklyn             Sheepshead Bay  40.586890 -73.943186
53        Brooklyn          Manhattan Terrace  40.614433 -73.957438
54        Brooklyn                   Flatbush  40.636326 -73.958401
55        Brooklyn              Crown Heights  40.670829 -73.943291
56        Brooklyn              East Flatbush  40.641718 -73.936103
57        Brooklyn                 Kensington  40.642382 -73.980421

In [42]:
CLIENT_ID = 'IGDNA3PDM4UQQ5FXGHSREYQPDKA2D42B4OQ5NETKJR51ZSKS' 
CLIENT_SECRET = 'SWY3FEM4CDYI0G5KK5N12DHZCELH15GRKKI3001ZSU3RN3BG' 
VERSION = '20180605'
LIMIT = 100 # 

print('Your credentails:')
print('CLIENT_ID: ' + CLIENT_ID)
print('CLIENT_SECRET:' + CLIENT_SECRET)

Your credentails:
CLIENT_ID: IGDNA3PDM4UQQ5FXGHSREYQPDKA2D42B4OQ5NETKJR51ZSKS
CLIENT_SECRET:SWY3FEM4CDYI0G5KK5N12DHZCELH15GRKKI3001ZSU3RN3BG


In [51]:
from bs4 import BeautifulSoup
import requests
import pandas as pd

Now Borough and Neighborhood Canada

List of postal codes of Canada

In [52]:
List_url = "https://en.wikipedia.org/wiki/List_of_postal_codes_of_Canada:_M"
source = requests.get(List_url).text

In [53]:
soup = BeautifulSoup(source, 'xml')

In [54]:
table=soup.find('table')

In the case of Canada, an additional filter is added which is the postal code.

In [55]:
column_names = ['Postalcode','Borough','Neighborhood']
df = pd.DataFrame(columns = column_names)

In [56]:
for tr_cell in table.find_all('tr'):
    row_data=[]
    for td_cell in tr_cell.find_all('td'):
        row_data.append(td_cell.text.strip())
    if len(row_data)==3:
        df.loc[len(df)] = row_data

In [57]:
df.head()

Postalcode           Borough               Neighborhood
0        M1A      Not assigned               Not assigned
1        M2A      Not assigned               Not assigned
2        M3A        North York                  Parkwoods
3        M4A        North York           Victoria Village
4        M5A  Downtown Toronto  Regent Park, Harbourfront

Data Cleaning

In [58]:
df=df[df['Borough']!='Not assigned']

In [59]:
df=df[df['Neighborhood']!='Not assigned']

In [60]:
df.head()

Postalcode           Borough                                 Neighborhood
2        M3A        North York                                    Parkwoods
3        M4A        North York                             Victoria Village
4        M5A  Downtown Toronto                    Regent Park, Harbourfront
5        M6A        North York             Lawrence Manor, Lawrence Heights
6        M7A  Downtown Toronto  Queen's Park, Ontario Provincial Government

In [61]:
temp_df=df.groupby('Postalcode')['Neighborhood'].apply(lambda x: "%s" % ', '.join(x))
temp_df=temp_df.reset_index(drop=False)
temp_df.rename(columns={'Neighborhood':'Neighborhood_joined'},inplace=True)

In [62]:
df_merge = pd.merge(df, temp_df, on='Postalcode')

In [63]:
df_merge.drop(['Neighborhood'],axis=1,inplace=True)

In [64]:
df_merge.drop_duplicates(inplace=True)

In [65]:
df_merge.rename(columns={'Neighborhood_joined':'Neighborhood'},inplace=True)

In [66]:
df_merge.head()

Postalcode           Borough                                 Neighborhood
0        M3A        North York                                    Parkwoods
1        M4A        North York                             Victoria Village
2        M5A  Downtown Toronto                    Regent Park, Harbourfront
3        M6A        North York             Lawrence Manor, Lawrence Heights
4        M7A  Downtown Toronto  Queen's Park, Ontario Provincial Government

In [67]:
df_merge.shape

(103, 3)

Once the preliminary analysis has been carried out and the filter is carried out by postal code with the coordinates of the sites

In [68]:
def get_geocode(postal_code):
   
    lat_lng_coords = None
    while(lat_lng_coords is None):
        g = geocoder.google('{}, Toronto, Ontario'.format(postal_code))
        lat_lng_coords = g.latlng
    latitude = lat_lng_coords[0]
    longitude = lat_lng_coords[1]
    return latitude,longitude

In [69]:
geo_df=pd.read_csv('http://cocl.us/Geospatial_data')

In [70]:
geo_df.head()

Postal Code   Latitude  Longitude
0         M1B  43.806686 -79.194353
1         M1C  43.784535 -79.160497
2         M1E  43.763573 -79.188711
3         M1G  43.770992 -79.216917
4         M1H  43.773136 -79.239476

In [71]:
geo_df

Postal Code   Latitude  Longitude
0           M1B  43.806686 -79.194353
1           M1C  43.784535 -79.160497
2           M1E  43.763573 -79.188711
3           M1G  43.770992 -79.216917
4           M1H  43.773136 -79.239476
5           M1J  43.744734 -79.239476
6           M1K  43.727929 -79.262029
7           M1L  43.711112 -79.284577
8           M1M  43.716316 -79.239476
9           M1N  43.692657 -79.264848
10          M1P  43.757410 -79.273304
11          M1R  43.750072 -79.295849
12          M1S  43.794200 -79.262029
13          M1T  43.781638 -79.304302
14          M1V  43.815252 -79.284577
15          M1W  43.799525 -79.318389
16          M1X  43.836125 -79.205636
17          M2H  43.803762 -79.363452
18          M2J  43.778517 -79.346556
19          M2K  43.786947 -79.385975
20          M2L  43.757490 -79.374714
21          M2M  43.789053 -79.408493
22          M2N  43.770120 -79.408493
23          M2P  43.752758 -79.400049
24          M2R  43.782736 -79.442259
25          M3A  43.753259 -79.329656
26          M3B  43.745906 -79.352188
27          M3C  43.725900 -79.340923
28          M3H  43.754328 -79.442259
29          M3J  43.767980 -79.487262
30          M3K  43.737473 -79.464763
31          M3L  43.739015 -79.506944
32          M3M  43.728496 -79.495697
33          M3N  43.761631 -79.520999
34          M4A  43.725882 -79.315572
35          M4B  43.706397 -79.309937
36          M4C  43.695344 -79.318389
37          M4E  43.676357 -79.293031
38          M4G  43.709060 -79.363452
39          M4H  43.705369 -79.349372
40          M4J  43.685347 -79.338106
41          M4K  43.679557 -79.352188
42          M4L  43.668999 -79.315572
43          M4M  43.659526 -79.340923
44          M4N  43.728020 -79.388790
45          M4P  43.712751 -79.390197
46          M4R  43.715383 -79.405678
47          M4S  43.704324 -79.388790
48          M4T  43.689574 -79.383160
49          M4V  43.686412 -79.400049
50          M4W  43.679563 -79.377529
51          M4X  43.667967 -79.367675
52          M4Y  43.665860 -79.383160
53          M5A  43.654260 -79.360636
54          M5B  43.657162 -79.378937
55          M5C  43.651494 -79.375418
56          M5E  43.644771 -79.373306
57          M5G  43.657952 -79.387383
58          M5H  43.650571 -79.384568
59          M5J  43.640816 -79.381752
60          M5K  43.647177 -79.381576
61          M5L  43.648198 -79.379817
62          M5M  43.733283 -79.419750
63          M5N  43.711695 -79.416936
64          M5P  43.696948 -79.411307
65          M5R  43.672710 -79.405678
66          M5S  43.662696 -79.400049
67          M5T  43.653206 -79.400049
68          M5V  43.628947 -79.394420
69          M5W  43.646435 -79.374846
70          M5X  43.648429 -79.382280
71          M6A  43.718518 -79.464763
72          M6B  43.709577 -79.445073
73          M6C  43.693781 -79.428191
74          M6E  43.689026 -79.453512
75          M6G  43.669542 -79.422564
76          M6H  43.669005 -79.442259
77          M6J  43.647927 -79.419750
78          M6K  43.636847 -79.428191
79          M6L  43.713756 -79.490074
80          M6M  43.691116 -79.476013
81          M6N  43.673185 -79.487262
82          M6P  43.661608 -79.464763
83          M6R  43.648960 -79.456325
84          M6S  43.651571 -79.484450
85          M7A  43.662301 -79.389494
86          M7R  43.636966 -79.615819
87          M7Y  43.662744 -79.321558
88          M8V  43.605647 -79.501321
89          M8W  43.602414 -79.543484
90          M8X  43.653654 -79.506944
91          M8Y  43.636258 -79.498509
92          M8Z  43.628841 -79.520999
93          M9A  43.667856 -79.532242
94          M9B  43.650943 -79.554724
95          M9C  43.643515 -79.577201
96          M9L  43.756303 -79.565963
97          M9M  43.724766 -79.532242
98          M9N  43.706876 -79.518188
99          M9P  43.696319 -79.532242
100         M9R  43.688905 -79.554724
101         M9V  43.739416 -79.588437
102         M9W  43.706748 -79.594054

After this, the postal code can be correlated with the coordinates of the sites, indicating city and neighborhood

In [72]:
geo_df.rename(columns={'Postal Code':'Postalcode'},inplace=True)
geo_merged = pd.merge(geo_df, df_merge, on='Postalcode')

In [73]:
geo_merged

Postalcode   Latitude  Longitude           Borough  \
0          M1B  43.806686 -79.194353       Scarborough   
1          M1C  43.784535 -79.160497       Scarborough   
2          M1E  43.763573 -79.188711       Scarborough   
3          M1G  43.770992 -79.216917       Scarborough   
4          M1H  43.773136 -79.239476       Scarborough   
5          M1J  43.744734 -79.239476       Scarborough   
6          M1K  43.727929 -79.262029       Scarborough   
7          M1L  43.711112 -79.284577       Scarborough   
8          M1M  43.716316 -79.239476       Scarborough   
9          M1N  43.692657 -79.264848       Scarborough   
10         M1P  43.757410 -79.273304       Scarborough   
11         M1R  43.750072 -79.295849       Scarborough   
12         M1S  43.794200 -79.262029       Scarborough   
13         M1T  43.781638 -79.304302       Scarborough   
14         M1V  43.815252 -79.284577       Scarborough   
15         M1W  43.799525 -79.318389       Scarborough   
16         M1X  43.836125 -79.205636       Scarborough   
17         M2H  43.803762 -79.363452        North York   
18         M2J  43.778517 -79.346556        North York   
19         M2K  43.786947 -79.385975        North York   
20         M2L  43.757490 -79.374714        North York   
21         M2M  43.789053 -79.408493        North York   
22         M2N  43.770120 -79.408493        North York   
23         M2P  43.752758 -79.400049        North York   
24         M2R  43.782736 -79.442259        North York   
25         M3A  43.753259 -79.329656        North York   
26         M3B  43.745906 -79.352188        North York   
27         M3C  43.725900 -79.340923        North York   
28         M3H  43.754328 -79.442259        North York   
29         M3J  43.767980 -79.487262        North York   
30         M3K  43.737473 -79.464763        North York   
31         M3L  43.739015 -79.506944        North York   
32         M3M  43.728496 -79.495697        North York   
33         M3N  43.761631 -79.520999        North York   
34         M4A  43.725882 -79.315572        North York   
35         M4B  43.706397 -79.309937         East York   
36         M4C  43.695344 -79.318389         East York   
37         M4E  43.676357 -79.293031      East Toronto   
38         M4G  43.709060 -79.363452         East York   
39         M4H  43.705369 -79.349372         East York   
40         M4J  43.685347 -79.338106         East York   
41         M4K  43.679557 -79.352188      East Toronto   
42         M4L  43.668999 -79.315572      East Toronto   
43         M4M  43.659526 -79.340923      East Toronto   
44         M4N  43.728020 -79.388790   Central Toronto   
45         M4P  43.712751 -79.390197   Central Toronto   
46         M4R  43.715383 -79.405678   Central Toronto   
47         M4S  43.704324 -79.388790   Central Toronto   
48         M4T  43.689574 -79.383160   Central Toronto   
49         M4V  43.686412 -79.400049   Central Toronto   
50         M4W  43.679563 -79.377529  Downtown Toronto   
51         M4X  43.667967 -79.367675  Downtown Toronto   
52         M4Y  43.665860 -79.383160  Downtown Toronto   
53         M5A  43.654260 -79.360636  Downtown Toronto   
54         M5B  43.657162 -79.378937  Downtown Toronto   
55         M5C  43.651494 -79.375418  Downtown Toronto   
56         M5E  43.644771 -79.373306  Downtown Toronto   
57         M5G  43.657952 -79.387383  Downtown Toronto   
58         M5H  43.650571 -79.384568  Downtown Toronto   
59         M5J  43.640816 -79.381752  Downtown Toronto   
60         M5K  43.647177 -79.381576  Downtown Toronto   
61         M5L  43.648198 -79.379817  Downtown Toronto   
62         M5M  43.733283 -79.419750        North York   
63         M5N  43.711695 -79.416936   Central Toronto   
64         M5P  43.696948 -79.411307   Central Toronto   
65         M5R  43.672710 -79.405678   Central Toronto   
66         M5S  43.662696 -79.400049  Downtown Toronto   
67         M5T  43.653206 -79.400049  Downtown Toronto   
68

In [74]:
geo_merged.head()

Postalcode   Latitude  Longitude      Borough  \
0        M1B  43.806686 -79.194353  Scarborough   
1        M1C  43.784535 -79.160497  Scarborough   
2        M1E  43.763573 -79.188711  Scarborough   
3        M1G  43.770992 -79.216917  Scarborough   
4        M1H  43.773136 -79.239476  Scarborough   

                             Neighborhood  
0                          Malvern, Rouge  
1  Rouge Hill, Port Union, Highland Creek  
2       Guildwood, Morningside, West Hill  
3                                  Woburn  
4                               Cedarbrae

Now that the information is complete we can filter by Borough

Filter--Toronto

In [75]:
toronto_data=geo_merged[geo_merged['Borough'].str.contains("Toronto")]
toronto_data.head()

Postalcode   Latitude  Longitude          Borough  \
37        M4E  43.676357 -79.293031     East Toronto   
41        M4K  43.679557 -79.352188     East Toronto   
42        M4L  43.668999 -79.315572     East Toronto   
43        M4M  43.659526 -79.340923     East Toronto   
44        M4N  43.728020 -79.388790  Central Toronto   

                      Neighborhood  
37                     The Beaches  
41    The Danforth West, Riverdale  
42  India Bazaar, The Beaches West  
43                 Studio District  
44                   Lawrence Park

Filter--York

In [79]:
York_data=geo_merged[geo_merged['Borough'].str.contains("York")]
York_data

Postalcode   Latitude  Longitude     Borough  \
17        M2H  43.803762 -79.363452  North York   
18        M2J  43.778517 -79.346556  North York   
19        M2K  43.786947 -79.385975  North York   
20        M2L  43.757490 -79.374714  North York   
21        M2M  43.789053 -79.408493  North York   
22        M2N  43.770120 -79.408493  North York   
23        M2P  43.752758 -79.400049  North York   
24        M2R  43.782736 -79.442259  North York   
25        M3A  43.753259 -79.329656  North York   
26        M3B  43.745906 -79.352188  North York   
27        M3C  43.725900 -79.340923  North York   
28        M3H  43.754328 -79.442259  North York   
29        M3J  43.767980 -79.487262  North York   
30        M3K  43.737473 -79.464763  North York   
31        M3L  43.739015 -79.506944  North York   
32        M3M  43.728496 -79.495697  North York   
33        M3N  43.761631 -79.520999  North York   
34        M4A  43.725882 -79.315572  North York   
35        M4B  43.706397 -79.309937   East York   
36        M4C  43.695344 -79.318389   East York   
38        M4G  43.709060 -79.363452   East York   
39        M4H  43.705369 -79.349372   East York   
40        M4J  43.685347 -79.338106   East York   
62        M5M  43.733283 -79.419750  North York   
71        M6A  43.718518 -79.464763  North York   
72        M6B  43.709577 -79.445073  North York   
73        M6C  43.693781 -79.428191        York   
74        M6E  43.689026 -79.453512        York   
79        M6L  43.713756 -79.490074  North York   
80        M6M  43.691116 -79.476013        York   
81        M6N  43.673185 -79.487262        York   
96        M9L  43.756303 -79.565963  North York   
97        M9M  43.724766 -79.532242  North York   
98        M9N  43.706876 -79.518188        York   

                                        Neighborhood  
17                                 Hillcrest Village  
18                      Fairview, Henry Farm, Oriole  
19                                   Bayview Village  
20                          York Mills, Silver Hills  
21                           Willowdale, Newtonbrook  
22                       Willowdale, Willowdale East  
23                                   York Mills West  
24                       Willowdale, Willowdale West  
25                                         Parkwoods  
26                                         Don Mills  
27                                         Don Mills  
28   Bathurst Manor, Wilson Heights, Downsview North  
29                   Northwood Park, York University  
30                                         Downsview  
31                                         Downsview  
32                                         Downsview  
33                                         Downsview  
34                                  Victoria Village  
35                   Parkview Hill, Woodbine Gardens  
36                                  Woodbine Heights  
38                                           Leaside  
39                                  Thorncliffe Park  
40     East Toronto, Broadview North (Old East York)  
62                 Bedford Park, Lawrence Manor East  
71                  Lawrence Manor, Lawrence Heights  
72                                         Glencairn  
73                                Humewood-Cedarvale  
74                               Caledonia-Fairbanks  
79          North Park, Maple Leaf Park, Upwood Park  
80  Del Ray, Mount Dennis, Keelsdale and Silverthorn  
81                     Runnymede, The Junction North  
96                                     Humber Summit  
97                                  Humberlea, Emery  
98                                            Weston

Filter--Scarborough

In [80]:
Scarborough_data=geo_merged[geo_merged['Borough'].str.contains("Scarborough")]
Scarborough_data

Postalcode   Latitude  Longitude      Borough  \
0         M1B  43.806686 -79.194353  Scarborough   
1         M1C  43.784535 -79.160497  Scarborough   
2         M1E  43.763573 -79.188711  Scarborough   
3         M1G  43.770992 -79.216917  Scarborough   
4         M1H  43.773136 -79.239476  Scarborough   
5         M1J  43.744734 -79.239476  Scarborough   
6         M1K  43.727929 -79.262029  Scarborough   
7         M1L  43.711112 -79.284577  Scarborough   
8         M1M  43.716316 -79.239476  Scarborough   
9         M1N  43.692657 -79.264848  Scarborough   
10        M1P  43.757410 -79.273304  Scarborough   
11        M1R  43.750072 -79.295849  Scarborough   
12        M1S  43.794200 -79.262029  Scarborough   
13        M1T  43.781638 -79.304302  Scarborough   
14        M1V  43.815252 -79.284577  Scarborough   
15        M1W  43.799525 -79.318389  Scarborough   
16        M1X  43.836125 -79.205636  Scarborough   

                                         Neighborhood  
0                                      Malvern, Rouge  
1              Rouge Hill, Port Union, Highland Creek  
2                   Guildwood, Morningside, West Hill  
3                                              Woburn  
4                                           Cedarbrae  
5                                 Scarborough Village  
6         Kennedy Park, Ionview, East Birchmount Park  
7                     Golden Mile, Clairlea, Oakridge  
8     Cliffside, Cliffcrest, Scarborough Village West  
9                         Birch Cliff, Cliffside West  
10  Dorset Park, Wexford Heights, Scarborough Town...  
11                                  Wexford, Maryvale  
12                                          Agincourt  
13            Clarks Corners, Tam O'Shanter, Sullivan  
14  Milliken, Agincourt North, Steeles East, L'Amo...  
15                      Steeles West, L'Amoreaux West  
16                                        Upper Rouge

In [81]:
def getNearbyVenues(names, latitudes, longitudes):
    radius=500
    LIMIT=100
    venues_list=[]
    for name, lat, lng in zip(names, latitudes, longitudes):
        print(name)
            
        # create the API request URL
        url = 'https://api.foursquare.com/v2/venues/explore?&client_id={}&client_secret={}&v={}&ll={},{}&radius={}&limit={}'.format(
            CLIENT_ID, 
            CLIENT_SECRET, 
            VERSION, 
            lat, 
            lng, 
            radius, 
            LIMIT)
            
        # make the GET request
        results = requests.get(url).json()["response"]['groups'][0]['items']
        
        # return only relevant information for each nearby venue
        venues_list.append([(
            name, 
            lat, 
            lng, 
            v['venue']['name'], 
            v['venue']['location']['lat'], 
            v['venue']['location']['lng'],  
            v['venue']['categories'][0]['name']) for v in results])

    nearby_venues = pd.DataFrame([item for venue_list in venues_list for item in venue_list])
    nearby_venues.columns = ['Neighborhood', 
                  'Neighborhood Latitude', 
                  'Neighborhood Longitude', 
                  'Venue', 
                  'Venue Latitude', 
                  'Venue Longitude', 
                  'Venue Category']
    
    return(nearby_venues)

Now we can get some venues

1. Toronto-Venues

In [82]:
toronto_venues = getNearbyVenues(names=toronto_data['Neighborhood'],
                                   latitudes=toronto_data['Latitude'],
                                   longitudes=toronto_data['Longitude']
                                  )

Hillcrest Village
Fairview, Henry Farm, Oriole
Bayview Village
York Mills, Silver Hills
Willowdale, Newtonbrook
Willowdale, Willowdale East
York Mills West
Willowdale, Willowdale West
Parkwoods
Don Mills
Don Mills
Bathurst Manor, Wilson Heights, Downsview North
Northwood Park, York University
Downsview
Downsview
Downsview
Downsview
Victoria Village
Parkview Hill, Woodbine Gardens
Woodbine Heights
Leaside
Thorncliffe Park
East Toronto, Broadview North (Old East York)
Bedford Park, Lawrence Manor East
Lawrence Manor, Lawrence Heights
Glencairn
Humewood-Cedarvale
Caledonia-Fairbanks
North Park, Maple Leaf Park, Upwood Park
Del Ray, Mount Dennis, Keelsdale and Silverthorn
Runnymede, The Junction North
Humber Summit
Humberlea, Emery
Weston


2. York-Venues

In [83]:
York_venues = getNearbyVenues(names=York_data['Neighborhood'],
                                   latitudes=York_data['Latitude'],
                                   longitudes=York_data['Longitude']
                                  )

Hillcrest Village
Fairview, Henry Farm, Oriole
Bayview Village
York Mills, Silver Hills
Willowdale, Newtonbrook
Willowdale, Willowdale East
York Mills West
Willowdale, Willowdale West
Parkwoods
Don Mills
Don Mills
Bathurst Manor, Wilson Heights, Downsview North
Northwood Park, York University
Downsview
Downsview
Downsview
Downsview
Victoria Village
Parkview Hill, Woodbine Gardens
Woodbine Heights
Leaside
Thorncliffe Park
East Toronto, Broadview North (Old East York)
Bedford Park, Lawrence Manor East
Lawrence Manor, Lawrence Heights
Glencairn
Humewood-Cedarvale
Caledonia-Fairbanks
North Park, Maple Leaf Park, Upwood Park
Del Ray, Mount Dennis, Keelsdale and Silverthorn
Runnymede, The Junction North
Humber Summit
Humberlea, Emery
Weston


3.Scarborough-Venues

In [84]:
Scarborough_venues = getNearbyVenues(names=Scarborough_data['Neighborhood'],
                                   latitudes=Scarborough_data['Latitude'],
                                   longitudes=Scarborough_data['Longitude']
                                  )

Malvern, Rouge
Rouge Hill, Port Union, Highland Creek
Guildwood, Morningside, West Hill
Woburn
Cedarbrae
Scarborough Village
Kennedy Park, Ionview, East Birchmount Park
Golden Mile, Clairlea, Oakridge
Cliffside, Cliffcrest, Scarborough Village West
Birch Cliff, Cliffside West
Dorset Park, Wexford Heights, Scarborough Town Centre
Wexford, Maryvale
Agincourt
Clarks Corners, Tam O'Shanter, Sullivan
Milliken, Agincourt North, Steeles East, L'Amoreaux East
Steeles West, L'Amoreaux West
Upper Rouge


Now we can see the information in tables

In [85]:
toronto_venues.head()

Neighborhood  Neighborhood Latitude  Neighborhood Longitude  \
0  Hillcrest Village              43.803762              -79.363452   
1  Hillcrest Village              43.803762              -79.363452   
2  Hillcrest Village              43.803762              -79.363452   
3  Hillcrest Village              43.803762              -79.363452   
4  Hillcrest Village              43.803762              -79.363452   

                                 Venue  Venue Latitude  Venue Longitude  \
0               Eagle's Nest Golf Club       43.805455       -79.364186   
1                      AY Jackson Pool       43.804515       -79.366138   
2                         Villa Madina       43.801685       -79.363938   
3                    Duncan Creek Park       43.805539       -79.360695   
4  A.Y. Jackson Secondary School Track       43.805068       -79.366677   

             Venue Category  
0               Golf Course  
1                      Pool  
2  Mediterranean Restaurant  
3                   Dog Run  
4        Athletics & Sports

In [86]:
York_venues.head()

Neighborhood  Neighborhood Latitude  Neighborhood Longitude  \
0  Hillcrest Village              43.803762              -79.363452   
1  Hillcrest Village              43.803762              -79.363452   
2  Hillcrest Village              43.803762              -79.363452   
3  Hillcrest Village              43.803762              -79.363452   
4  Hillcrest Village              43.803762              -79.363452   

                                 Venue  Venue Latitude  Venue Longitude  \
0               Eagle's Nest Golf Club       43.805455       -79.364186   
1                      AY Jackson Pool       43.804515       -79.366138   
2                         Villa Madina       43.801685       -79.363938   
3                    Duncan Creek Park       43.805539       -79.360695   
4  A.Y. Jackson Secondary School Track       43.805068       -79.366677   

             Venue Category  
0               Golf Course  
1                      Pool  
2  Mediterranean Restaurant  
3                   Dog Run  
4        Athletics & Sports

In [87]:
Scarborough_venues.head()

Neighborhood  Neighborhood Latitude  \
0                          Malvern, Rouge              43.806686   
1  Rouge Hill, Port Union, Highland Creek              43.784535   
2  Rouge Hill, Port Union, Highland Creek              43.784535   
3       Guildwood, Morningside, West Hill              43.763573   
4       Guildwood, Morningside, West Hill              43.763573   

   Neighborhood Longitude                                              Venue  \
0              -79.194353                                            Wendy’s   
1              -79.160497                              Royal Canadian Legion   
2              -79.160497  SEBS Engineering Inc. (Sustainable Energy and ...   
3              -79.188711                                     RBC Royal Bank   
4              -79.188711                                  G & G Electronics   

   Venue Latitude  Venue Longitude              Venue Category  
0       43.807448       -79.199056        Fast Food Restaurant  
1       43.782533       -79.163085                         Bar  
2       43.782371       -79.156820  Construction & Landscaping  
3       43.766790       -79.191151                        Bank  
4       43.765309       -79.191537           Electronics Store

RESULTS AND CONCLUSIONS

* With the available information it is possible to analyze some aspects as it was carried out, in the second case of the Canadian information, an information filtering was carried out with the help of the Foursquare API tool.
*The management of the information and the filtering of it is important to be able to carry out the different analyzes, it is very important once the information is available to convert it to a data frame and to be able to perform the respective filters whatever is required.